### PR TITLE
Better handling of unassigned topologies and other mismatches with st…

### PR DESCRIPTION
**Pull request by bergloman**
…orage

### CODE REVIEW

File: src/distributed/topology_worker.ts [hunk 1]
```diff
@@ -99,6 +94,46 @@ export class TopologyWorker {
         this.coordinator.run();
     }
 
+    /** Thi smethod verifies tha t all topologies are running and properly registered */
```
Comment (rupnikj) [inline src/distributed/topology_worker.ts]: typos

File: src/distributed/topology_leader.ts [hunk 2]
```diff
@@ -203,53 +207,77 @@ export class TopologyLeader {
                     if (!perform_loop || alive_workers.length == 0) {
```
Comment (rupnikj) [inline src/distributed/topology_leader.ts]: line 194: if (alive_workers.length == 0)
Why do we handle dead worksers if at least one is live and not handle otherwise?

File: src/distributed/topology_leader.ts [hunk 3]
```diff
@@ -203,53 +207,77 @@ export class TopologyLeader {
...
+                (xcallback) => {
+                    if (!perform_loop || alive_workers.length == 0) {
+                        return xcallback();
+                    }
+                    let targets = topologies_enabled
+                        .filter(x => x.status == intf.Consts.TopologyStatus.running && x.worker == null);
+                    async.each(
+                        targets,
+                        (item: intf.TopologyStatus, xxcallback) => {
+                            // strange, topology marked as enabled and running, but no worker specified
```
Comment (rupnikj) [inline src/distributed/topology_leader.ts]: what happens when it gets assigned later since it's already running?

File: src/distributed/topology_leader.ts [hunk 4]
```diff
@@ -203,53 +207,77 @@ export class TopologyLeader {
...
+                            // mark it as unassinged.                        
+                            log.logger().important(this.log_prefix + "Topology marked as running and enabled, but no worker specified: " + item.uuid);
+                            self.storage.setTopologyStatus(item.uuid, intf.Consts.TopologyStatus.unassigned, null, xxcallback);
+                        },
+                        xcallback
+                    );
+                },
+                (xcallback) => {
```
Comment (rupnikj) [inline src/distributed/topology_leader.ts]: some comments on various stages would be nice, it's hard to follow what's being done in each callback

File: src/distributed/topology_worker.ts [hunk 5]
```diff
@@ -99,6 +94,46 @@ export class TopologyWorker {
...
+    private verifyTopologies(uuids: string[], callback: intf.SimpleCallback) {
```
Comment (rupnikj) [inline src/distributed/topology_worker.ts]: verifyTopologies does more than verify. A change of name to indicate that for start would be good. Reporting inconsistencies not just fixing would also be good.

File: src/distributed/topology_worker.ts (hunk 1 above)
Comment (bergloman) [reply src/distributed/topology_worker.ts]: fixed

File: src/distributed/topology_leader.ts (hunk 2 above)
Comment (bergloman) [reply src/distributed/topology_leader.ts]: I agree. fixed.

File: src/distributed/topology_leader.ts (hunk 3 above)
Comment (bergloman) [reply src/distributed/topology_leader.ts]: It means that worker didn't properly set the status to unassigned. If we don't know where it is running, set it as unassigned.
The worker process on will kill it when the verify function will be called.

File: src/distributed/topology_leader.ts (hunk 4 above)
Comment (bergloman) [reply src/distributed/topology_leader.ts]: Working on it.

File: src/distributed/topology_worker.ts (hunk 5 above)
Comment (bergloman) [reply src/distributed/topology_worker.ts]: I agree. Will rename it.
